src/variable.jl

```diff
@@ -310,82 +310,61 @@ nomissing(a::AbstractArray,value) = a
-    @debug "$(@__LINE__)"
+function readblock!(v::Variable{T,N}, aout, indexes::TR...) where {T,N} where TR <: Union{StepRange{Int,Int},UnitRange{Int}}
+    start,count,stride,jlshape = ncsub(indexes[1:N])
+    data = Array{T,N}(undef,jlshape)
     datamode(v.ds)
-    # use zero-based indexes and reversed order
-    nc_put_var1(v.ds.ncid,v.varid,[i-1 for i in indexes[ndims(v):-1:1]],T(data))
-    return data
+    nc_get_vars!(v.ds.ncid,v.varid,start,count,stride,data)
+    # nc_get_vars!(v.ds.ncid,v.varid,start,count,stride,aout)
+    aout .= data
+    return aout
 end
 
-function Base.getindex(v::Variable{T,N},indexes::Colon...) where {T,N}
-    datamode(v.ds)
-    data = Array{T,N}(undef,size(v))
-    nc_get_var!(v.ds.ncid,v.varid,data)
+function readblock!(v::Variable{T,N}, aout, indexes::Union{Int,Colon,AbstractRange{<:Integer}}...) where {T,N}
+    sz = size(v)
+    start,count,stride = ncsub2(sz,indexes...)
+    jlshape = _shape_after_slice(sz,indexes...)
+    data = Array{T}(undef,jlshape)
 
-    # special case for scalar NetCDF variable
-    if N == 0
-        return data[]
-    else
-        return data
-    end
+    datamode(v.ds)
+    nc_get_vars!(v.ds.ncid,v.varid,start,count,stride,data)
+    aout .= data
 end
 
-function Base.setindex!(v::Variable{T,N},data::T,indexes::Colon...) where {T,N}
-    @debug "setindex! colon $data"
-    datamode(v.ds) # make sure that the file is in data mode
-    tmp = fill(data,size(v))
-    nc_put_var(v.ds.ncid,v.varid,tmp)
+# NetCDF scalars indexed as []
+readblock!(v::Variable{T, 0}, aout) where T = readblock!(v, aout, 1)
+# writeblock!(v::Variable{T, 0}, data) where T = writeblock!(v, data, 1)
+
+function writeblock!(v::Variable{T,N},data,indexes::AbstractUnitRange...) where {T,N}
+    datamode(v.ds)
+    _write_data_to_nc(v, data, indexes...)
     return data
 end
 
-# union types cannot be used to avoid ambiguity
-for data_type = [Number, String, Char]
-    @eval begin
-        # call to v .= 123
-        function Base.setindex!(v::Variable{T,N},data::$data_type) where {T,N}
-            @debug "setindex! $data"
-            datamode(v.ds) # make sure that the file is in data mode
-            tmp = fill(convert(T,data),size(v))
-            nc_put_var(v.ds.ncid,v.varid,tmp)
-            return data
-        end
-
-        Base.setindex!(v::Variable,data::$data_type,indexes::Colon...) = setindex!(v::Variable,data)
-
-        function Base.setindex!(v::Variable{T,N},data::$data_type,indexes::StepRange{Int,Int}...) where {T,N}
-            datamode(v.ds) # make sure that the file is in data mode
-            start,count,stride,jlshape = ncsub(indexes[1:ndims(v)])
-            tmp = fill(convert(T,data),jlshape)
-            nc_put_vars(v.ds.ncid,v.varid,start,count,stride,tmp)
-            return data
-        end
-    end
+function _write_data_to_nc(v::Variable{T,N},data,indexes::Int...) where {T,N}
+    nc_put_var1(v.ds.ncid,v.varid,[i-1 for i in indexes[ndims(v):-1:1]],T(data[1]))
```

Comment:
```suggestion
    nc_put_var1(v.ds.ncid,v.varid,[i-1 for i in reverse(indexes),T(data[1]))
```

DiskArrays.jl should take care of the `1:ndims` part? if not you would want to check those are actually trailing ones.